exposed-jdbc/src/main/kotlin/org/jetbrains/exposed/sql/statements/jdbc/JdbcPreparedStatementImpl.kt

```diff
@@ -38,7 +43,25 @@ class JdbcPreparedStatementImpl(val statement: PreparedStatement, val wasGenerat
     }
 
     override fun setInputStream(index: Int, inputStream: InputStream) {
-        statement.setBinaryStream(index, inputStream, inputStream.available())
+        try {
+            when {
+                // streams with known length where available matches the actual length
+                inputStream is ByteArrayInputStream ->
+                    statement.setBinaryStream(index, inputStream, inputStream.available())
+
+                // FileInputStream.available() returns returns Int.MAX_VALUE
+                // if the underlying file is larger than 2GB
+                inputStream is FileInputStream && inputStream.available() < Int.MAX_VALUE ->
+                    statement.setBinaryStream(index, inputStream, inputStream.available())
+
+                // default handling for unknown length
+                else -> statement.setBinaryStream(index, inputStream)
+
+            }
+        } catch (e: SQLFeatureNotSupportedException) {
+            // fallback to bytes
+            statement.setBytes(index, inputStream.readAllBytes())
```

Comment:
I tested it and it appears that this line works for all cases. Why not use just this?

Reply:
`readAllBytes()` will fail on contents larger than 2GB (maximum size of the ByteArray) and allocates the whole data within RAM. The `setBinaryStream` function transfers just small chunks to the database until the stream is fully read (and allocates only one of those small chunks on the heap).

Reply:
Thanks for clarifying.

One thing I'm concerned about is having to check the type of `InputStream` and do something specific for each type. `InputStream` has a very large number of sub-classes and I wonder how sustainable/maintainable it is to have a separate approach for each one instead of having a generic solution that could potentially work for most, if not all of them, if that's possible.

Also, I tried testing it using a 3GB file created using RandomAccessFile and it failed for all databases for being too large.
```
        val fileSize = 3L * 1024L * 1024L * 1024L // 3GB
        val filePath = "path_to_large_file.txt"

        try {
            val randomAccessFile = RandomAccessFile(filePath, "rw")
            randomAccessFile.setLength(fileSize)
            println("Large file created: $filePath")
            randomAccessFile.close()
        } catch (e: IOException) {
            e.printStackTrace()
        }
        
        val longBlob = ExposedBlob(
              inputStream = FileInputStream(filePath)
        )
```
Is your solution supposed to cover this case?

Reply:
Unfortunately, there is no interface within JDK to decide if the `InputStream` has a known size. But maybe, it's easier to skip the functions with the provided length.

I'll have to look into your 3GB test-case, it should have worked, maybe, it is necessary to use `Connection.createBlob` instead of `InputStream`. I hoped, this would be transparently handled by the JDBC-driver.

Reply:
I'll split the PR into two separate PRs: one for the `available()` bug (with is more urgent for me) and the blob handling with `Connection.createBlob()` (which currently fails on columns with default values).